optimum/onnxruntime/optimization.py

```diff
@@ -0,0 +1,274 @@
+#  Copyright 2021 The HuggingFace Team. All rights reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import logging
+import os
+from pathlib import Path
+from typing import Callable, Optional, Union
+from collections import OrderedDict
+
+from transformers import AutoTokenizer
+from transformers.onnx import export
+from transformers.onnx.features import FeaturesManager
+
+import onnx
+from onnx import load_model
+from onnxruntime.transformers.optimizer import optimize_model, get_fusion_statistics, FusionOptions
+from onnxruntime.transformers.onnx_model_bert import BertOnnxModel
+from optimum.onnxruntime.configuration import ORTConfig
+from optimum.onnxruntime.utils import generate_identified_filename
+
+
+logger = logging.getLogger(__name__)
+
+class AttrDict(dict):
+    def __init__(self, *args, **kwargs):
+        super(AttrDict, self).__init__(*args, **kwargs)
+        self.__dict__ = self
+
+
+class OnnxConfigManager:
+    """A class that notes down the attribute names(for `number of heads` and `hidden size`) of models in `transformers/models`. 
+    It is optional for BERT model, but for other model types, you need specify the name of these parameters. It is possible to add
+    customized model information with `update_model()` method.
+        Args:
+            __conf (:obj:`dict`):
+                Dictionary register the attribute names of models.
+        """
+    __conf = {
+        "bert": {"num_heads":"num_attention_heads", "hidden_size":"hidden_size"},
+        "distilbert": {"num_heads":"n_heads", "hidden_size":"hidden_size"},
+        "roberta": {"num_heads":"num_attention_heads", "hidden_size":"hidden_size"},
+        "bart": {"num_heads":"encoder_attention_heads", "hidden_size":"d_model"},
+        "gpt2": {"num_heads":"n_head", "hidden_size":"n_embd"},
+        }
```

Comment:
As there is a redundancy in the dictionnaries keys ("num_heads" and "hidden_size"), we could potentially replace the dictionnaries with tuples

Reply:
Good idea, it shall look lighter with tuples.

Reply:
I am not sure that tuples are better, it is clearer to read `OnnxConfigManager.__conf[model_type]["num_heads"]` than `OnnxConfigManager.__conf[model_type][0]` in my opinion,